addons/osfstorage/models.py

```diff
@@ -399,10 +416,11 @@ def is_checked_out(self):
 
     @property
     def is_preprint_primary(self):
-        if self.node.preprint_file:
-            for child in self.children.filter(node__preprint_file=self.node.preprint_file).select_related('node'):
-                if child.is_preprint_primary:
-                    return True
+        if hasattr(self.target, 'preprint_file') and self.target.preprint_file:
+            for child in self.children.all():
```

Comment:
This looks costly: it's doing 1-2 queries for each child (for accessing `.target`), *recursively*. IIRC, generic foreign keys can be prefetched, which should save a lot of queries. Can you try doing `self.children.all().prefetch_related('target')`?